packages/@aws-cdk/custom-resources/lib/provider-framework/state-machine.ts

```diff
@@ -0,0 +1,80 @@
+import { Grant, IGrantable, PolicyStatement, Role, ServicePrincipal } from '@aws-cdk/aws-iam';
+import { IFunction } from '@aws-cdk/aws-lambda';
+import { CfnResource, Construct, Duration } from '@aws-cdk/core';
+
+export interface StateMachineProps {
+  readonly isCompleteHandler: IFunction;
+
+  readonly timeoutHandler: IFunction;
+
+  readonly interval: Duration;
+
+  readonly maxAttempts: number;
+
+  readonly backoffRate: number;
+}
+
+/**
+ * A very simple StateMachine construct highly customized to the provider framework.
+ * This is so that this package does not need to depend on aws-stepfunctions module.
+ */
+export class StateMachine extends Construct {
+  public readonly stateMachineArn: string;
+
+  constructor(scope: Construct, id: string, props: StateMachineProps) {
+    super(scope, id);
+
+    const role = new Role(this, 'Role', {
+      assumedBy: new ServicePrincipal('states.amazonaws.com'),
+    });
+    role.addToPolicy(new PolicyStatement({
+      actions: [ 'lambda:InvokeFunction' ],
+      resources: [ props.isCompleteHandler.functionArn, props.timeoutHandler.functionArn ],
+    }));
+
+    const definition = JSON.stringify({
```

Comment:
I wonder if it makes more sense to use [Stack.toJsonString](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_core.Stack.html#to-wbr-json-wbr-stringobj-space) instead so that tokens are properly handled.

Reply:
Nice! I wasn't aware of this.